baseapp/streaming.go

```diff
@@ -0,0 +1,105 @@
+package baseapp
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/spf13/cast"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	servertypes "github.com/cosmos/cosmos-sdk/server/types"
+	store "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// ABCIListener is the interface that we're exposing as a streaming service.
+// It hooks into the ABCI message processing of the BaseApp.
+// The error results are propagated to consensus state machine,
+// if you don't want to affect consensus, handle the errors internally and always return `nil` in these APIs.
+type ABCIListener interface {
+	// ListenBeginBlock updates the streaming service with the latest BeginBlock messages
+	ListenBeginBlock(ctx context.Context, req abci.RequestBeginBlock, res abci.ResponseBeginBlock) error
+	// ListenEndBlock updates the steaming service with the latest EndBlock messages
+	ListenEndBlock(ctx context.Context, req abci.RequestEndBlock, res abci.ResponseEndBlock) error
+	// ListenDeliverTx updates the steaming service with the latest DeliverTx messages
+	ListenDeliverTx(ctx context.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) error
+	// ListenCommit updates the steaming service with the latest Commit messages and state changes
+	ListenCommit(ctx context.Context, res abci.ResponseCommit, changeSet []*store.StoreKVPair) error
+}
+
+const (
+	StreamingTomlKey                  = "streaming"
+	StreamingABCITomlKey              = "abci"
+	StreamingABCIPluginTomlKey        = "plugin"
+	StreamingABCIKeysTomlKey          = "keys"
+	StreamingABCIStopNodeOnErrTomlKey = "stop-node-on-err"
+	StreamingABCIAsync                = "async"
+)
+
+// RegisterStreamingPlugin registers streaming plugins with the App.
+func RegisterStreamingPlugin(
+	bApp *BaseApp,
+	appOpts servertypes.AppOptions,
+	keys map[string]*store.KVStoreKey,
+	streamingPlugin interface{},
+) error {
+	switch t := streamingPlugin.(type) {
+	case ABCIListener:
+		registerABCIListenerPlugin(bApp, appOpts, keys, t)
+	default:
+		return fmt.Errorf("unexpected plugin type %T", t)
+	}
+	return nil
+}
+
+func registerABCIListenerPlugin(
+	bApp *BaseApp,
+	appOpts servertypes.AppOptions,
+	keys map[string]*store.KVStoreKey,
+	abciListener ABCIListener,
+) {
+	asyncKey := fmt.Sprintf("%s.%s.%s", StreamingTomlKey, StreamingABCITomlKey, StreamingABCIAsync)
+	async := cast.ToBool(appOpts.Get(asyncKey))
+	stopNodeOnErrKey := fmt.Sprintf("%s.%s.%s", StreamingTomlKey, StreamingABCITomlKey, StreamingABCIStopNodeOnErrTomlKey)
+	stopNodeOnErr := cast.ToBool(appOpts.Get(stopNodeOnErrKey))
+	keysKey := fmt.Sprintf("%s.%s.%s", StreamingTomlKey, StreamingABCITomlKey, StreamingABCIKeysTomlKey)
+	exposeKeysStr := cast.ToStringSlice(appOpts.Get(keysKey))
+	exposedKeys := exposeStoreKeysSorted(exposeKeysStr, keys)
+	bApp.cms.AddListeners(exposedKeys)
+	bApp.SetStreamingService(abciListener)
+	bApp.stopNodeOnABCIListenerErr = stopNodeOnErr
```

Comment:
I am skeptical that BaseApp is the best place for `stopNodeOnABCIListenerErr`, since it leads to the kind of [question you raised](https://github.com/cosmos/cosmos-sdk/pull/14207/files#r1043726514).  

I propose passing this configuration state into the listener instead of BaseApp.  What do you think?

Reply:
I'm resolving this conversation. It continues in https://github.com/cosmos/cosmos-sdk/pull/14207#discussion_r1067564955